### PR TITLE
Port some more DDA magic code

### PR DESCRIPTION
**Pull request by olanti-p**
#### Purpose of change
1. Fix `line_attack` spells with `aoe <= 1` ignore walls and don't damage creatures in their path
2. Add ability for player to learn spells by installing bionics
3. Add tests for magic/spells

#### Describe the solution
All code is cherry-picked from DDA repo with almost no modifications.
CleverRaven#39940 (and fix for tests from #40637)
CleverRaven#40242
CleverRaven#40328

#### Testing
Tests pass; manually tested the bionic.

### CODE REVIEW

File: src/bionics.cpp [hunk 1]
```diff
@@ -2398,6 +2399,28 @@ void Character::add_bionic( const bionic_id &b )
         add_bionic( inc_bid );
     }
 
+    for( const std::pair<spell_id, int> spell_pair : b->learned_spells ) {
```
Comment (Coolthulhu) [inline src/bionics.cpp:2402]:
```
error: loop variable 'spell_pair' has type 'const std::pair<spell_id, int> &' (aka 'const pair<string_id<spell_type>, int> &') but is initialized with type 'const std::pair<const string_id<spell_type>, int>' resulting in a copy [-Werror,-Wrange-loop-construct]
```

File: src/bionics.cpp [hunk 2]
```diff
@@ -2418,8 +2443,20 @@ void Character::remove_bionic( const bionic_id &b )
...
 
+        for( const std::pair<spell_id, int> &spell_pair : i.id->learned_spells ) {
+            cbm_spells.emplace( spell_pair.first );
+        }
+
         new_my_bionics.push_back( bionic( i.id, i.invlet ) );
     }
+
+    // any spells you learn from installing a bionic you forget.
+    for( const std::pair<spell_id, int> &spell_pair : b->learned_spells ) {
```
Comment (Coolthulhu) [inline src/bionics.cpp:2454]: Clang complains about non-const `spell_id` in the pair.

File: src/bionics.cpp [hunk 3]
```diff
@@ -2418,8 +2443,20 @@ void Character::remove_bionic( const bionic_id &b )
             continue;
         }
 
+        for( const std::pair<spell_id, int> &spell_pair : i.id->learned_spells ) {
```
Comment (Coolthulhu) [inline src/bionics.cpp:2446]: Clang complains about non-const `spell_id` in the pair.

**Comment by olanti-p** [comment]
Should be fixed

**Comment by Coolthulhu** [comment]
I didn't realize spellcasting is unbound by default

**Pull request merged**